src/application/notifications.ts

```diff
@@ -31,23 +36,45 @@ export class NotificationsService {
     this.emailClient.setAuth(this.emailAuth);
   }
 
-  public async notifySuccess(recipient: User): Promise<void> {
+  public async notifyError(
+    recipient: User,
+    repoCanonicalName: string,
+    tag: string,
+    errors: Error[]
+  ): Promise<void> {
     await this.setAuth();
-    await this.emailClient.sendEmail(
-      recipient.email,
-      this.sender,
-      "Publish to BCR [no-reply]",
-      "Successfully published pull request to the Bazel Central Registry."
-    );
-  }
 
-  public async notifyError(recipient: User): Promise<void> {
-    await this.setAuth();
+    const subject = `Publish to BCR [${repoCanonicalName}]`;
+
+    let content = `\
+Failed to publish entry for ${repoCanonicalName}@${tag} to the Bazel Central Registry.
+
+`;
+
+    const friendlyErrors = errors.filter(isUserFacingError);
+
+    for (let error of friendlyErrors) {
+      content += `${error.message}\n\n`;
+    }
+
+    if (!friendlyErrors.length) {
+      content +=
+        "An unknown error occurred. Please report an issue here: https://github.com/bazel-contrib/publish-to-bcr/issues.";
+    }
+
     await this.emailClient.sendEmail(
       recipient.email,
       this.sender,
-      "Publish to BCR [no-reply]",
-      "Failed"
+      subject,
+      content
     );
   }
 }
+
+function isUserFacingError(error: Error) {
+  return (
+    error instanceof MissingFilesError ||
```

Comment:
can the information about isUserFacing live inside those error types rather than out here? even if they have to carry a runtime value it seems better per the locality principle

Reply:
Yeah good idea.